File: App/Clients/ng-microservices-demo-client/src/app/infrastructure/rest-clients/policies/models/dtos/question-answer-dto.ts
import { AutoMap } from "@automapper/classes";
import { QuestionType } from "../../../../../application/enums/question-type";

export abstract class QuestionAnswerDto {
  @AutoMap()
  questionCode!: string;

  abstract get questionType(): QuestionType;
  abstract getAnswer(): unknown;
}

export abstract class QuestionAnswerDtoGeneric<T> extends QuestionAnswerDto {
  @AutoMap()
  answer!: T;

  override getAnswer(): T {
    return this.answer;
  }
}

File: App/Services/Microservices.Demo.Pricing.Service/Application/UseCases/Tariff/Dtos/QuestionAnswerDto.cs
using Microservices.Demo.Pricing.Service.Application.UseCases.Tariff.Enums;

namespace Microservices.Demo.Pricing.Service.Application.UseCases.Tariff.Dtos
{
    public abstract class QuestionAnswerDto
    {
        public string QuestionCode { get; set; }
        public abstract QuestionType QuestionType { get; }
        public abstract object GetAnswer();
    }
    public abstract class QuestionAnswerModel<T> : QuestionAnswerDto
    {
        public T Answer { get; set; }

        public override object GetAnswer()
        {
            return Answer;
        }
    }
}
